### PR TITLE
fix iou precision error

### PR DESCRIPTION
**Pull request by h-bo**
currently, when cal iou, yolox use float16 (in float16 mode), but float16 only has 10 bits for valid part this may lead iou precision error:
for example torch.tensor(841, dtype=float16) -  torch.tensor(5.2, dtype=float16) ==torch.tensor(836., dtype=torch.float16). 

this can lead iou to be larger than 1 and in cls loss, yolox multiplies target with iou, so that the **bce loss can be negtive!**

change iou calculation to float32 seems to fix that
image
image

### CODE REVIEW

**Comment by FateScript** [comment]
Thanks for pointing it out. Another way to solve this issue is  using clip to ensure the iou value. Why not using it? @h-bo

**Comment by h-bo** [comment]
> Thanks for pointing it out. Another way to solve this issue is using clip to ensure the iou value. Why not using it? @h-bo

That's also a solution, I'm not sure which one is better. I just tried float32 in my dataset and it improves 2%~

**Comment by Joker316701882** [comment]
@h-bo Thanks! We will try it on COCO and see what happens.

**Comment by Joker316701882** [comment]
@h-bo Hi. We have tested this PR on COCO. Here are our reproduced results:
Before -> After
Small: 40.7->40.3
Medium: 46.9 -> 47.0
Large: 49.3 -> 49.3

Seems this PR does not affect the result on COCO. We recommend using ''value clip'' to tackle your reported issue. What do you think?

**Comment by h-bo** [comment]
@Joker316701882 It will be fine. Maybe the negtive cls loss problem is more severe in my dataset, if have time, I will dive into the reason. 
By the way, did you find the negative cls loss occurrence on coco?

**Comment by Joker316701882** [comment]
@h-bo 
I did not check this very carefully, but the results can tell that it is not a severe issue on COCO. Welcome to report your updated understanding of this issue!
